mlflow/langchain/databricks_dependencies.py

```diff
@@ -0,0 +1,185 @@
+        dependency_dict = _extract_dependency_dict_from_lc_model(
+            lc_model.retriever, dependency_dict
+        )
+    if hasattr(
+        lc_model, "llm_chain"
+    ):  # StuffDocumentsChain, MapRerankDocumentsChain, MapReduceDocumentsChain
+        dependency_dict = _extract_dependency_dict_from_lc_model(
+            lc_model.llm_chain.llm, dependency_dict
+        )
+    if hasattr(lc_model, "question_generator"):  # BaseConversationalRetrievalChain
+        dependency_dict = _extract_dependency_dict_from_lc_model(
+            lc_model.question_generator.llm, dependency_dict
+        )
+    if hasattr(lc_model, "initial_llm_chain") and hasattr(
+        lc_model, "refine_llm_chain"
+    ):  # RefineDocumentsChain
+        dependency_dict = _extract_dependency_dict_from_lc_model(
+            lc_model.initial_llm_chain.llm, dependency_dict
+        )
+        dependency_dict = _extract_dependency_dict_from_lc_model(
+            lc_model.refine_llm_chain.llm, dependency_dict
+        )
+
+    if hasattr(lc_model, "combine_documents_chain"):  # RetrievalQA, ReduceDocumentsChain
+        dependency_dict = _extract_dependency_dict_from_lc_model(
+            lc_model.combine_documents_chain, dependency_dict
+        )
+    if hasattr(lc_model, "combine_docs_chain"):  # BaseConversationalRetrievalChain
+        dependency_dict = _extract_dependency_dict_from_lc_model(
+            lc_model.combine_docs_chain, dependency_dict
+        )
+    if (
+        hasattr(lc_model, "collapse_documents_chain")
+        and lc_model.collapse_documents_chain is not None
+    ):  # ReduceDocumentsChain
+        dependency_dict = _extract_dependency_dict_from_lc_model(
+            lc_model.collapse_documents_chain, dependency_dict
+        )
+
+    return dependency_dict
+
+
+def _traverse_runnable(
+    lc_model, dependency_dict: Dict[str, str], visited: Set[str]
+) -> (Dict[str, str], Set[str]):
+    """
+    This function contains the logic to traverse a langchain_core.runnables.RunnableSerializable
+    object. It first inspects the current object using _extract_dependency_dict_from_lc_model
+    and then, if the current object is a Runnable, it recursively inspects its children returned
+    by lc_model.get_graph().nodes.values().
+    This function supports arbitrary LCEL chain.
+    """
+    import langchain_core
+
+    current_object_id = id(lc_model)
+    if current_object_id in visited:
+        return dependency_dict, visited
+
+    # Visit the current object
+    visited.add(current_object_id)
+    dependency_dict = _extract_dependency_dict_from_lc_model(lc_model, dependency_dict)
+
+    if isinstance(lc_model, langchain_core.runnables.RunnableSerializable):
+        # Visit the returned graph
+        for node in lc_model.get_graph().nodes.values():
+            dependency_dict, visited = _traverse_runnable(node.data, dependency_dict, visited)
+    else:
+        # This is a leaf node
+        pass
```

Comment:
Curious what's the lc_model type if it reaches here? BTW why does it have to be RunnableSerializable not Runnable?

Reply:
> what's the lc_model type if it reaches here?
I think it could be PromptTemplate, Input, OutputParser or things like that.

> why does it have to be RunnableSerializable not Runnable?
Probably it should be Runnable? I used RunnableSerializable because I see a Chain class inherits RunnableSerializable, so I just used it. Do you think Runnable is better? Happy to change to that if so!

Reply:
Actually PromptTemplate and Input/OutputParsers are also runnables 🤣 But I think it will end if there's no children nodes, so it should be fine. And yes I think Runnable is more straightforward :)

Reply:
sg! I'll use runnables here.

Shall we keep this no-op branch here in case there are something non-Runnable found (I can update the comment to say this instead of "leaf node"), or would you recommend removing this check?

Reply:
Yes keeping it here and updating the comment sounds good!